### PR TITLE
fix: Fix parsing of weapons with gear skins

### PR DESCRIPTION
**Pull request by P0keDev**
Weapons with skins would fail to parse if Wynncraft updated their skin at the same time a spell was cast - it would be detected as a new item, but not as a gear item because it would have the spell cast line as its name. On any further updates it would be detected as the same item and the annotation would never get updated.

### CODE REVIEW

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (magicus) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: I've tried hard to keep "specific" item knowledge out of the ItemHandler. Before you merge this, let me check a bit closer if there is a way to solve this which does not mean that the ItemHandler has to know about gear and skins...

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (magicus) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: This is a very handy shortcut. Why did you remove this? Did any of item type, damage or count change when applying a skin?

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (magicus) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: ~~I'm kind of thinking loud here, but what if you just turned `LoreUtils.isLoreEquals` into a `LoreUtils.isLoreEqualsIgnoringSkin`? Would that not solve the entire problem? And that means that the "knowledge" of the skin line is kind of contained in the LoreUtils function. And all it needs to do is filter out `^\\[Active .* Skin: \\]$` from both lore line lists before comparing.~~

~~It could be made even more general by turning it into like `isLoreEqualsIgnoringBracketLines` and just filter out `^\\[.*\\]$`, but maybe that is too generic for no good reason.~~

edit: Bad idea, ignore

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (magicus) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: Hm, I realize now that this line is strictly not necessary. What the code below checks is really "is the new lore identical to the old, except that it has an added line". That is a more general approach.

Sorry for all the confusing comments in random order here, but I think this approach is actually quite good, but you can drop the "isGearSkin" test and just dive directly into comparing the lore. You can also start by checking not only size inequality, but if new is exactly one line longer than existing, otherwise abort quickly.

And name the method not `sameGearWithSkin` but like `isLoreAlmostEquals` (and maybe move it to LoreUtils?)

Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: ... and then you can actually replace the original call to `isLoreEquals` with `isLoreAlmostEquals`, you don't need to check this twice.

Comment (P0keDev) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: This is a great point actually! I don't know off the top of my head if there are any other instances where lore would get "added to" like this, but it's probably safe to assume that if they exist, they can be handled in the same way. I will try to implement this later today

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (P0keDev) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: The item has to fundamentally change for a skin to work, so yes. I'm unsure if there's cases where the item itself will change or if it's only the item damage, but I'd say it's safest to remove the check altogether for this context

Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: The `similarStack` methods looks for three things: if the Item of the two ItemStacks are different. If they are, they can not be the same item. If they are the same, then the count is checked. If the count is different, it cannot be the same item. And finally, if the item and count matches, it also checks the "damage" value. If it is different, it cannot be the same value.

I think I need to log in and check weapon skins myself, but I have a hard time imagining that any of these change when you apply a skin. Or, oh, hm... does the damange value change maybe? The item looks differently in the hotbar with a skin applied?

I need to log in and experiment a bit.

Anway, it's kind of sad if we need to remove this. It's a quick shortcut that makes us skip reading and comparing lore texts for most item changes. Like each time you switch pages in the bank, then 99% of all items get caught in this check and we don't have to do the expensive lore reading.

At the very least we should be able to keep the count check. I hope we can also keep the item type, but we might have to skip the damage check.

Comment (P0keDev) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: weapon skins change the model of the weapon; that can't be achieved without at the very least changing the item's damage value

you're right about the stack count though, that can definitely be kept regardless

Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: Ok, I have checked a bit. As far as I could tell, *all* weapon skins are `diamond_shovel`. And head-wear skins are `diamond_pickaxe`. That means we could accept that these two items are "wildcard" items that match anything. So in similarStack, the short-circuit logic will return false if the two Items are different, but not if one of the two is a "wildcard" item.

something like this:
```
    private boolean similarStack(ItemStack firstItem, ItemStack secondItem) {
        if (!firstItem.getItem().equals(secondItem.getItem()) &&!isWildcardItem(firstItem) && !isWildcardItem(secondItem)) return false;
       
```
and `isWildcardItem` checks for equality with `Items.DIAMOND_PICKAXE` or `Items.DIAMOND_SHOVEL`.

Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java:91]: I realize this hard-codes a bit Wynncraft-y behavior, but I'm willing to sacrifice that to keep the high-performing short-circuit logic of `similarStack`.

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: Yes, if this kind of behavior would appear anywhere else, I think it will be reasonable to consider it the "same" item.

File: common/src/main/java/com/wynntils/utils/mc/LoreUtils.java
Comment (magicus) [inline common/src/main/java/com/wynntils/utils/mc/LoreUtils.java]: I'm thinking this should be symmetric, so it also returns true if the first item contains all of the second items lines.

Then it will catch the case of disabling a weapon skin as well.

Or maybe I'm missing something in my thinking here.

**Review by magicus — COMMENTED** [review]
Overall I think this looks very good. And it works, I hope? :-)

While this solves the transition "no skin" -> "has skin", I think the opposite would be nice as well, since the weapon you left will have its skin removed, and it would be nice to not have to re-parse the GearItem for that.

File: common/src/main/java/com/wynntils/utils/mc/LoreUtils.java
Comment (P0keDev) [reply common/src/main/java/com/wynntils/utils/mc/LoreUtils.java]: It shouldn't be possible for a skin getting disabled to cause any conflicts, since it can't coincide with a spell cast - but I might as well make this method symmetric anyway

Comment (magicus) [reply common/src/main/java/com/wynntils/utils/mc/LoreUtils.java]: It will save us a re-parse in GearAnnotator when you switch weapon. So yeah, it's not a bug fix, but a (ridiculous) micro-optimization.

**Review by magicus — APPROVED** [review]

**Pull request merged**
